src/libcore/slice.rs

```diff
@@ -295,22 +295,23 @@ impl<T> SliceExt for [T] {
     fn binary_search_by<F>(&self, mut f: F) -> Result<usize, usize> where
         F: FnMut(&T) -> Ordering
     {
-        let mut base : usize = 0;
-        let mut lim : usize = self.len();
+        let mut base = 0usize;
+        let mut s = self;
 
-        while lim != 0 {
-            let ix = base + (lim >> 1);
-            match f(&self[ix]) {
-                Equal => return Ok(ix),
+        loop {
+            let (head, tail) = s.split_at(s.len() >> 1);
+            if tail.is_empty() {
+                return Err(base)
+            }
+            match f(&tail[0]) {
+                Equal => return Ok(base + head.len()),
```

Comment:
I intentionally swapped `Equal` and `Greater`, in order to ensure that if the compiler is unable to fold all of the checks in a single comparison, the one for equality is only performed once in the whole search.